File: App/Resources/AlgorithmDetails/stacklessrotatemergesort/stacklessrotatemergesort.rb
def multi_swap(array, a, b, len)
  len.times do |i|
    array[a + i], array[b + i] = array[b + i], array[a + i]
  end
end

def rotate(array, a, m, b)
  l = m - a
  r = b - m
  while l > 0 && r > 0
    if r < l
      multi_swap(array, m - r, m, r)
      b -= r
      m -= r
      l -= r
    else
      multi_swap(array, a, m, l)
      a += l
      m += l
      r -= l
    end
  end
end

# Selects the c smallest combined elements of the two already-sorted runs
# array[a...m] and array[m...b] into the front half via a single rotation.
# Uses a merge-path (co-rank) binary search over whichever run is shorter:
# it looks for the split count r such that taking r elements from the tail
# of one run and (c - r) from the head of the other yields exactly the c
# smallest values in order, rather than searching for a value directly.
def partition_merge(array, a, m, b, c)
  len_a = m - a
  len_b = b - m
  return if len_a < 1 || len_b < 1

  if len_b < len_a
    cc = (len_a + len_b) - c
    r1 = [0, cc - len_a].max
    r2 = [cc, len_b].min
    while r1 < r2
      ml = r1 + (r2 - r1) / 2
      if array[m - (cc - ml)] > array[b - ml - 1]
        r2 = ml
      else
        r1 = ml + 1
      end
    end
    rotate(array, m - (cc - r1), m, b - r1)
  else
    r1 = [0, c - len_b].max
    r2 = [c, len_a].min
    while r1 < r2
      ml = r1 + (r2 - r1) / 2
      if array[a + ml] > array[m + (c - ml) - 1]
        r2 = ml
      else
        r1 = ml + 1
      end
    end
    rotate(array, a + r1, m, m + (c - r1))
  end
end

# Finds the first place inside array[a...b] where ascending order breaks,
# then partition-merges the sorted piece before it with the sorted piece
# after it. A no-op if array[a...b] is already one ascending run.
def rotate_merge(array, a, b, c)
  i = a + 1
  i += 1 while i < b && array[i - 1] <= array[i]
  partition_merge(array, a, i, b, c) if i < b
end

def rotate_partition_merge_sort(array, n)
  return if n < 2

  (1...n).step(2) do |i|
    array[i - 1], array[i] = array[i], array[i - 1] if array[i - 1] > array[i]
  end

  j = 2
  while j < n
    b1 = 0
    block_start = 0
    while block_start + j < n
      b1 = [block_start + 2 * j, n].min
      partition_merge(array, block_start, block_start + j, b1, j)
      block_start += 2 * j
    end

    k = j / 2
    while k > 1
      seam_start = 0
      while seam_start + k < b1
        seam_end = [seam_start + 2 * k, n].min
        rotate_merge(array, seam_start, seam_end, k)
        seam_start += 2 * k
      end
      k /= 2
    end

    (1...b1).step(2) do |m|
      array[m - 1], array[m] = array[m], array[m - 1] if array[m - 1] > array[m]
    end

    j *= 2
  end
end

def sort(arr)
  rotate_partition_merge_sort(arr, arr.length)
  arr
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/stacklessrotatemergesort/stacklessrotatemergesort.java
import java.util.Arrays;

public class stacklessrotatemergesort {
  private static void multiSwap(int[] arr, int a, int b, int len) {
    for (int i = 0; i < len; i++) {
      int t = arr[a + i];
      arr[a + i] = arr[b + i];
      arr[b + i] = t;
    }
  }

  private static void rotate(int[] arr, int a, int m, int b) {
    int l = m - a;
    int r = b - m;
    while (l > 0 && r > 0) {
      if (r < l) {
        multiSwap(arr, m - r, m, r);
        b -= r;
        m -= r;
        l -= r;
      } else {
        multiSwap(arr, a, m, l);
        a += l;
        m += l;
        r -= l;
      }
    }
  }

  /**
   * Selects the c smallest combined elements of the two already-sorted runs arr[a, m) and arr[m, b)
   * into the front half via a single rotation. Uses a merge-path (co-rank) binary search over
   * whichever run is shorter: it looks for the split count r such that taking r elements from the
   * tail of one run and (c - r) from the head of the other yields exactly the c smallest values in
   * order, rather than searching for a value directly.
   */
  private static void partitionMerge(int[] arr, int a, int m, int b, int c) {
    int lenA = m - a;
    int lenB = b - m;
    if (lenA < 1 || lenB < 1) {
      return;
    }

    if (lenB < lenA) {
      int cc = (lenA + lenB) - c;
      int r1 = Math.max(0, cc - lenA);
      int r2 = Math.min(cc, lenB);
      while (r1 < r2) {
        int ml = r1 + (r2 - r1) / 2;
        if (arr[m - (cc - ml)] > arr[b - ml - 1]) {
          r2 = ml;
        } else {
          r1 = ml + 1;
        }
      }
      rotate(arr, m - (cc - r1), m, b - r1);
    } else {
      int r1 = Math.max(0, c - lenB);
      int r2 = Math.min(c, lenA);
      while (r1 < r2) {
        int ml = r1 + (r2 - r1) / 2;
        if (arr[a + ml] > arr[m + (c - ml) - 1]) {
          r2 = ml;
        } else {
          r1 = ml + 1;
        }
      }
      rotate(arr, a + r1, m, m + (c - r1));
    }
  }

  /**
   * Finds the first place inside arr[a, b) where ascending order breaks, then partition-merges the
   * sorted piece before it with the sorted piece after it. A no-op if arr[a, b) is already one
   * ascending run.
   */
  private static void rotateMerge(int[] arr, int a, int b, int c) {
    int i = a + 1;
    while (i < b && arr[i - 1] <= arr[i]) {
      i++;
    }
    if (i < b) {
      partitionMerge(arr, a, i, b, c);
    }
  }

  private static void rotatePartitionMergeSort(int[] arr, int n) {
    if (n < 2) {
      return;
    }

    for (int i = 1; i < n; i += 2) {
      if (arr[i - 1] > arr[i]) {
        int t = arr[i - 1];
        arr[i - 1] = arr[i];
        arr[i] = t;
      }
    }

    for (int j = 2; j < n; j *= 2) {
      int b1 = 0;
      int blockStart = 0;
      while (blockStart + j < n) {
        b1 = Math.min(blockStart + 2 * j, n);
        partitionMerge(arr, blockStart, blockStart + j, b1, j);
        blockStart += 2 * j;
      }

      for (int k = j / 2; k > 1; k /= 2) {
        int seamStart = 0;
        while (seamStart + k < b1) {
          int seamEnd = Math.min(seamStart + 2 * k, n);
          rotateMerge(arr, seamStart, seamEnd, k);
          seamStart += 2 * k;
        }
      }

      for (int m = 1; m < b1; m += 2) {
        if (arr[m - 1] > arr[m]) {
          int t = arr[m - 1];
          arr[m - 1] = arr[m];
          arr[m] = t;
        }
      }
    }
  }

  public static void sort(int[] arr) {
    rotatePartitionMergeSort(arr, arr.length);
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
